Makefile.macros

```diff
@@ -0,0 +1,121 @@
+# Define a library and its build rules, register unit
+# tests for it.
+#
+# Basic usage :
+#   libfoo_SOURCES = a.cpp b.cpp ...
+#   libfoo_HEADERS = a.h b.h ...
+#   $(eval $(call common_library,libfoo))
+#
+# additional attributes supported (define before calling the macro):
+#   - libfoo_TESTS:
+#     list of tests for this library; each must compile to a binary
+#     by itself, and for each test there must be a single source file
+#   - libfoo_TEST_LIBS:
+#     extra libraries built in this project linked into each test
+#     (just specify the name without extension)
+#   - libfoo_TEST_EXTRA_LDFLAGS:
+#     additional LDFLAGS to be passed to link each test binary;
+#     use this to pull in external libraries (not built in this project)
+#
+# Depending on configuration builds only static or both
+# static and dynamic library. Depending on configuration
+# may also set up rules for test coverage report.
+define common_library
+
+### library build rules
+
+# add to the global list of sources and headers
+
+SOURCES += $$($1_SOURCES)
+HEADERS += $$($1_HEADERS)
+
+# rule to build static library
+
+$$(BUILD_OUT_PREFIX)$1.a: $$(patsubst %.cpp, $$(BUILD_OUT_PREFIX)%.la, $$($1_SOURCES))
+STATIC_LIBRARIES += $$(BUILD_OUT_PREFIX)$1.a
+
+# rule to build build elf shared library
+
+$$(BUILD_OUT_PREFIX)$1.so: $$(patsubst %.cpp, $$(BUILD_OUT_PREFIX)%.lo, $$($1_SOURCES))
+ifeq ($$(ENABLE_SHARED), yes)
+SHARED_LIBRARIES += $$(BUILD_OUT_PREFIX)$1.so
+endif
+
+### test rules
+
+# list of test binaries within the target build directory
+$1_TESTBINARIES = $$(patsubst %, $$(BUILD_OUT_PREFIX)%, $$($1_TESTS))
+
+# add to global list of tests
+TESTS += $$(patsubst %, $$(BUILD_OUT_PREFIX)%, $$($1_TESTS))
+
+# linking of test binaries
+$$($1_TESTBINARIES): $$(patsubst %, $$(BUILD_OUT_PREFIX)%.a, $$($1_TEST_LIBS)) $$(BUILD_OUT_PREFIX)$1.a
+$$($1_TESTBINARIES): LDFLAGS+=$$(patsubst %, $$(BUILD_OUT_PREFIX)%.a, $$($1_TEST_LIBS)) $$(BUILD_OUT_PREFIX)$1.a $$($1_TEST_EXTRA_LDFLAGS)
```

Comment:
This is annoying to resolve. The problem is that you would "normally" want the library under test to come first, such that it can resolve everything against the additional libraries required in the test. The problem is that jlmtest depends on libjlm, so there the ordering is wrong and has to be the other way round.

The only way to resolve is to remove implicit dependency on the library-under-test and need it mentioned explicitly as testlibs, so that is what I changed now.